cluster-agent/controllers/managed-gitops/eventloop/operation_event_loop.go

```diff
@@ -319,14 +426,174 @@ func (task *processOperationEventTask) internalPerformTask(taskContext context.C
+// returns shouldRetry, error
+func processOperation_SyncOperation(ctx context.Context, dbOperation db.Operation, crOperation operation.Operation, dbQueries db.DatabaseQueries,
+	argoCDNamespace corev1.Namespace, eventClient client.Client, credentialService *utils.CredentialService, log logr.Logger) (bool, error) {
+
+	// Sanity checks
+	if dbOperation.Resource_id == "" {
+		return shouldRetryFalse, fmt.Errorf("resource id was nil while processing operation: " + crOperation.Name)
+	}
+
+	// 1) Retrieve the SyncOperation DB entry pointed to by the Operation DB entry
+	dbSyncOperation := &db.SyncOperation{
+		SyncOperation_id: dbOperation.Resource_id,
+	}
+	if err := dbQueries.GetSyncOperationById(ctx, dbSyncOperation); err != nil {
+
+		if !db.IsResultNotFoundError(err) {
+			// On generic error, we should retry.
+			log.Error(err, "DB error occurred on retriving SyncOperation: "+dbSyncOperation.SyncOperation_id)
+			return shouldRetryTrue, err
+		} else {
+			// On db row not found, just ignore it and move on.
+			log.V(sharedutil.LogLevel_Debug).Info("SyncOperation '" + dbSyncOperation.SyncOperation_id + "' DB entry was no longer available.")
+			return shouldRetryFalse, err
+		}
+	}
+
+	// 2) Retrieve the Application DB entry pointed to by the SyncOperation DB entry
+	dbApplication := db.Application{
+		Application_id: dbSyncOperation.Application_id,
+	}
+	if err := dbQueries.GetApplicationById(ctx, &dbApplication); err != nil {
+
+		if db.IsResultNotFoundError(err) {
+			log.Error(err, "Unable to retrieve application ID in SyncOperation table: "+dbApplication.Application_id)
+			return shouldRetryFalse, err
+		} else {
+			// On generic error, return true so the operation is retried.
+			log.Error(err, "Error occurred on retrieving application ID in SyncOperation table")
+			return shouldRetryTrue, err
+		}
+	}
+
+	// 3) Process the event, based on whether the SyncOperation is requesting an app sync, or a terminate.
+	if dbSyncOperation.DesiredState == db.SyncOperation_DesiredState_Running {
+
+		return runAppSync(ctx, dbOperation, *dbSyncOperation, &dbApplication, argoCDNamespace, credentialService, dbQueries, eventClient, log)
+
+	} else if dbSyncOperation.DesiredState == db.SyncOperation_DesiredState_Terminated {
+
+		return terminateExistingOperation(ctx, &dbApplication, argoCDNamespace, credentialService, eventClient, log)
+
+	} else {
+		log.Error(nil, "SEVERE: unexpected desired state in SyncOperation DB entry")
+		return shouldRetryFalse, nil
+	}
+}
+
+// returns shouldRetry, error
+func terminateExistingOperation(ctx context.Context, dbApplication *db.Application, argoCDNamespace corev1.Namespace,
+	credentialService *utils.CredentialService, eventClient client.Client, log logr.Logger) (bool, error) {
+
+	if err := utils.TerminateOperation(ctx, dbApplication.Name, argoCDNamespace, credentialService, eventClient, time.Duration(5*time.Minute), log); err != nil {
```

Comment:
what happens in a situation where this doesn't get deleted (for whatever reason) ? Given that we process on operation at a time, would that create a block the queue for the rest?

Reply:
Processing of each operation runs in its own goroutine, sharded by the resource that the operation is pointing to. See `taskRetryLoop.AddTaskIfNotPresent(...)` in this file for where the goroutine is started.

So for example, in this case the following operations would be processed as follows
Operation A: points to Application row A1
Operation B: points to Application row A2
Operation C: points to Application row A1
Operation D: points to SyncOperation row SO1
Operation E: points to SyncOperation row SO1
Operation F: points to SyncOperation row SO2

B, and F would run on their own goroutines, in parallel.
A/C, and D/E would would run sequentially (e.g. first A, then C).

Reply:
all clear now, thx